SU2_CFD/include/numerics/turbulent/transition/trans_sources.hpp

```diff
@@ -0,0 +1,258 @@
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+#include "../../../../../Common/include/toolboxes/geometry_toolbox.hpp"
+#include "../../scalar/scalar_sources.hpp"
+
+
+/*!
+ * \class CSourcePieceWise_TranLM
+ * \brief Class for integrating the source terms of the LM transition model equations.
+ * \ingroup SourceDiscr
+ * \author S. Kang.
+ */
+template <class FlowIndices>
+class CSourcePieceWise_TransLM final : public CNumerics {
+ private:
+  const FlowIndices idx; /*!< \brief Object to manage the access to the flow primitives. */
+
+  su2double g_eff_i,
+  g_eff_j,
+  g_sep_i,
+  g_sep_j;
+  /*--- LM Closure constants ---*/
+  const su2double c_e1 = 1.0;
+  const su2double c_a1 = 2.0;
+  const su2double c_e2 = 50.0;
+  const su2double c_a2 = 0.06;
+  const su2double sigmaf = 1.0;
+  const su2double s1 = 2.0;
+  const su2double c_theta = 0.03;
+  const su2double sigmat = 2.0;
+
+  su2double Vorticity;
+  su2double Residual[2];
+  su2double* Jacobian_i[2];
+  su2double Jacobian_Buffer[4];// Static storage for the Jacobian (which needs to be pointer for return type).
+
+ public:
+  /*!
+   * \brief Constructor of the class.
+   * \param[in] val_nDim - Number of dimensions of the problem.
+   * \param[in] val_nVar - Number of variables of the problem.
+   * \param[in] config - Definition of the particular problem.
+   */
+  CSourcePieceWise_TransLM(unsigned short val_nDim, unsigned short val_nVar, const CConfig* config) 
+      : CNumerics(val_nDim, 2, config),
+        idx(val_nDim, config->GetnSpecies()) {
+    
+    /*--- "Allocate" the Jacobian using the static buffer. ---*/
+    Jacobian_i[0] = Jacobian_Buffer;
+    Jacobian_i[1] = Jacobian_Buffer + 2;
+  }
+
+  /*!
+   * \brief Residual for source term integration.
+   * \param[in] config - Definition of the particular problem.
+   * \return A lightweight const-view (read-only) of the residual/flux and Jacobians.
+   */
+  ResidualType<> ComputeResidual(const CConfig* config) override {
+  /*--- ScalarVar[0] = k, ScalarVar[0] = w, TransVar[0] = gamma, and TransVar[0] = ReThetaT ---*/
+  /*--- dU/dx = PrimVar_Grad[1][0] ---*/
+  AD::StartPreacc();
+  AD::SetPreaccIn(StrainMag_i);
+  AD::SetPreaccIn(ScalarVar_i, nVar);
+  AD::SetPreaccIn(ScalarVar_Grad_i, nVar, nDim);
+  AD::SetPreaccIn(TransVar_i, nVar);
+  AD::SetPreaccIn(TransVar_Grad_i, nVar, nDim);
+  AD::SetPreaccIn(Volume); AD::SetPreaccIn(dist_i);
+  AD::SetPreaccIn(&V_i[idx.Velocity()], nDim);
+  AD::SetPreaccIn(PrimVar_Grad_i, nDim+idx.Velocity(), nDim);
```

Comment:
Why do we add nDim to idx.Velocity here?

Reply:
I used the preaccumulation function by referring to another hpp file. The AD::SetPreaccIn(&V_i[idx.Velocity()], nDim) is copied from scala_convection.hpp. I'm not sure what role the nDim in AD::SetPreaccIn(&V_i[idx.Velocity()], nDim) plays, so it was used same. So, no special reason for the nDim.

Would I modify it as below?

```suggestion
  AD::SetPreaccIn(PrimVar_Grad_i, nDim+idx.Velocity());
```

Reply:
What you have is correct, no changes are needed.
PrimVar_Grad_i is a matrix, we are registering the velocity gradients, which start at idx.Velocity, so we cover e.g. a 4x3 matrix

Reply:
Ah I apologize, I was asking out of ignorance.  Thanks for the clarification @pcarruscag.

Reply:
Oh... I'm sorry. 
I made a wrong suggestion.
not "AD::SetPreaccIn(PrimVar_Grad_i, nDim+idx.Velocity(), nDim);"




  "AD::SetPreaccIn(&V_i[idx.Velocity()], nDim); " is the issue.